src/rpc/server.ts

```diff
@@ -898,4 +902,103 @@ export class Server {
+   *    for assets are specific to a network. You can refer to {@link Networks}
+   *    for a list of built-in passphrases, e.g., `Networks.TESTNET`.
+   *
+   * @returns {Promise<Api.ContractBalanceResponse>}, which will contain the
+   *    trustline details if and only if the request returned a valid balance
+   *    ledger entry. If it doesn't, the `trustline` field will not exist.
+   *
+   * @throws {TypeError} If `contractId` is not a valid contract strkey (C...).
+   *
+   * @warning This should not be used for fetching custom token contracts, only
+   *    SACs. Using them with custom tokens is a security concern because they
+   *    can format their balance entries in any way they want, and thus this
+   *    fetch can be very misleading.
+   *
+   * @see getLedgerEntries
+   */
+  public async getContractBalance(
+    contractId: string,
+    token: Asset,
+    networkPassphrase?: string
+  ): Promise<Api.ContractBalanceResponse> {
+      if (!StrKey.isValidContract(contractId)) {
+        throw new TypeError(`expected contract ID, got ${contractId}`);
+      }
+
+      // Call out to RPC if passphrase isn't provided.
+      const passphrase: string = networkPassphrase
+        ?? await this.getNetwork().then(n => n.passphrase);
+
+      // Turn token into predictable contract ID
+      const tokenId = token.contractId(passphrase);
+
+      // Rust union enum type with "Balance(ScAddress)" structure
+      const key = xdr.ScVal.scvVec([
+        nativeToScVal("Balance", { type: "symbol" }),
+        nativeToScVal(contractId, { type: "address" }),
+      ]);
+
+      // Note a quirk here: the contract address in the key is the *token*
+      // rather than the *holding contract*. This is because each token stores a
+      // balance entry for each contract, not the other way around (i.e. XLM
+      // holds a reserve for contract X, rather that contract X having a balance
+      // of N XLM).
+      const ledgerKey = xdr.LedgerKey.contractData(
+        new xdr.LedgerKeyContractData({
+          contract: new Address(tokenId).toScAddress(),
+          durability: xdr.ContractDataDurability.persistent(),
+          key
+        })
+      );
+
+      const response = await this.getLedgerEntries(ledgerKey);
+      if (response.entries.length === 0) {
+        return { latestLedger: response.latestLedger };
+      }
+
+      const {
+        lastModifiedLedgerSeq,
+        liveUntilLedgerSeq,
+        val
+      } = response.entries[0];
+
+      if (val.switch().value !== xdr.LedgerEntryType.contractData().value) {
+        return { latestLedger: response.latestLedger };
+      }
+
+      // If any field doesn't match *exactly* what we expect, we bail. This
+      // prevents confusion with "balance-like" entries (e.g., has `amount` but
+      // isn't a bigint), but still allows "looks like a duck" balance entries.
+      const entry = scValToNative(val.contractData().val());
+      if (typeof entry.amount === 'bigint' &&
```

Comment:
I think this condition is unnecessary. Contract IDs for SAC are built using the underlying asset, and you're querying from an SAC, so if data is returned for the Balance key from the SAC, it must follow the structure we expect.

Having said that, I'm making the assumption that the `Asset` type here can only represent non-custom assets. I also might be unaware of some quirks in the server that might require this check, so let me know.

Reply:
That's a fantastic point; I was being overtly cautious since I wanted to avoid all possibility of returning an invalid balance entry. However, you're absolutely right that by requiring the user to pass in an `Asset`, it must in fact correspond to a real SAC (which may or may not exist yet, but that's beside the point since there'd just be no entry). I've removed a buncha code in f527390d per this comment; thank you!

Reply:
Btw honestly your :heavy_check_mark: is all I'd need to be comfortable merging this.